Modules/_io/textio.c

```diff
@@ -1498,15 +1498,26 @@ textiowrapper_read_chunk(textio *self, Py_ssize_t size_hint)
         /* Given this, we know there was a valid snapshot point
          * len(dec_buffer) bytes ago with decoder state (b'', dec_flags).
          */
-        if (PyArg_ParseTuple(state, "OO", &dec_buffer, &dec_flags) < 0) {
+        if (!PyTuple_Check(state)) {
+            PyErr_Format(PyExc_TypeError,
+                         "decoder getstate() should have returned a tuple, "
+                         "not '%.200s'",
+                         Py_TYPE(state)->tp_name);
+            Py_DECREF(state);
+            return -1;
+        }
+        if (!PyArg_ParseTuple(state,
+                              "OO;decoder getstate() should have returned a "
+                              "2-tuple", &dec_buffer, &dec_flags))
+        {
             Py_DECREF(state);
             return -1;
         }
 
         if (!PyBytes_Check(dec_buffer)) {
             PyErr_Format(PyExc_TypeError,
-                         "decoder getstate() should have returned a bytes "
-                         "object, not '%.200s'",
+                         "decoder getstate() should have returned a tuple "
```

Comment:
I thought it would be more helpful for users, but if you think this is an overkill, I would change it.
BTW, there is another `"decoder getstate() should have returned ..."` error message, in `DECODER_GETSTATE()`. should I change it also to "illegal decoder state" for consistency?
(the PyArg_ParseTuple before that one is part of bpo-28261. I am working on a patch for it, and I guessed we shouldn't fix it here.)